File: frontend/src/services/chatService.ts
import { v4 as uuidv4 } from 'uuid';
import type { ChatMessage } from '../types';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

export interface ChatCompletionRequest {
  model: string;
  messages: ChatMessage[];
  temperature?: number;
  top_p?: number;
  max_tokens?: number;
  stream?: boolean;
}

export interface ChatCompletionResponse {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: {
    index: number;
    message: ChatMessage;
    finish_reason: string;
  }[];
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

export interface SessionResponse {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

export interface MessageResponse {
  id: string;
  session_id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  created_at: string;
  tokens: number;
  metadata: Record<string, any> | null;
}

class ChatService {
  // 创建新会话
  async createSession(title: string): Promise<SessionResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ title }),
      });
      
      if (!response.ok) {
        throw new Error(`创建会话失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('创建会话时发生错误:', error);
      throw error;
    }
  }

  // 获取会话列表
  async getSessions(): Promise<SessionResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions`);
      
      if (!response.ok) {
        throw new Error(`获取会话列表失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('获取会话列表时发生错误:', error);
      throw error;
    }
  }

  // 获取会话详情
  async getSession(sessionId: string): Promise<SessionResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}`);
      
      if (!response.ok) {
        throw new Error(`获取会话详情失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('获取会话详情时发生错误:', error);
      throw error;
    }
  }

  // 更新会话
  async updateSession(sessionId: string, title: string): Promise<SessionResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ title }),
      });
      
      if (!response.ok) {
        throw new Error(`更新会话失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('更新会话时发生错误:', error);
      throw error;
    }
  }

  // 删除会话
  async deleteSession(sessionId: string): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}`, {
        method: 'DELETE',
      });
      
      if (!response.ok) {
        throw new Error(`删除会话失败: ${response.statusText}`);
      }
    } catch (error) {
      console.error('删除会话时发生错误:', error);
      throw error;
    }
  }

  // 添加消息到会话
  async addMessage(sessionId: string, message: Omit<MessageResponse, 'id' | 'session_id' | 'created_at'>): Promise<MessageResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}/messages`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(message),
      });
      
      if (!response.ok) {
        throw new Error(`添加消息失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('添加消息时发生错误:', error);
      throw error;
    }
  }

  // 获取会话消息
  async getMessages(sessionId: string): Promise<MessageResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}/messages`);
      
      if (!response.ok) {
        throw new Error(`获取消息失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('获取消息时发生错误:', error);
      throw error;
    }
  }

  // 创建聊天完成
  async createChatCompletion(request: ChatCompletionRequest, sessionId?: string): Promise<ChatCompletionResponse> {
    try {
      const url = sessionId 
        ? `${API_BASE_URL}/chat/completions?session_id=${sessionId}`
        : `${API_BASE_URL}/chat/completions`;
      
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
      });
      
      if (!response.ok) {
        throw new Error(`聊天完成处理失败: ${response.statusText}`);
      }
      
      return response.json();
    } catch (error) {
      console.error('聊天完成处理时发生错误:', error);
      throw error;
    }
  }
  
  // 生成唯一ID
  generateId(prefix: string = ''): string {
    return `${prefix}${uuidv4()}`;
  }
}

export default new ChatService();

File: backend/app/services/chat_service.py
from fastapi import HTTPException
import asyncio
import json
import time
from app.models.inference import LLMInference
from app.models.schemas import (
    ChatCompletionRequest, 
    ChatCompletionResponse, 
    ChatCompletionChoice, 
    ChatCompletionUsage,
    ChatMessage,
    ChatCompletionChunk,
    ChatCompletionChunkChoice,
    ChatCompletionChunkDelta
)
from app.services.database_service import DatabaseService
from app.utils import get_logger, generate_id, format_timestamp

logger = get_logger()
llm = LLMInference()
database_service = DatabaseService()


class ChatService:
    """聊天服务类"""
    
    async def generate_completion(self, request: ChatCompletionRequest, session_id: str = None, db = None) -> ChatCompletionResponse:
        """生成聊天完成响应"""
        try:
            # 记录开始时间
            start_time = time.time()
            
            # 生成回复
            response_text = llm.chat(
                messages=request.messages,
                max_new_tokens=request.max_tokens or 150,
                temperature=request.temperature,
                top_p=request.top_p
            )
            
            # 计算耗时
            elapsed_time = time.time() - start_time
            
            # 计算token数量（简化计算）
            prompt_tokens = sum(len(msg.content) if isinstance(msg.content, str) else sum(len(item.text) for item in msg.content if item.type == "text" and item.text) for msg in request.messages)
            completion_tokens = len(response_text)
            total_tokens = prompt_tokens + completion_tokens
            
            # 如果提供了会话ID且有数据库连接，将消息保存到数据库
            if session_id and db:
                # 保存用户消息
                for msg in request.messages:
                    database_service.add_message(
                        db, 
                        session_id, 
                        msg.role, 
                        msg.content, 
                        len(msg.content) if isinstance(msg.content, str) else 0
                    )
                
                # 保存AI回复
                database_service.add_message(
                    db, 
                    session_id, 
                    "assistant", 
                    response_text, 
                    completion_tokens
                )
            
            # 构造响应
            response = ChatCompletionResponse(
                id=generate_id("chatcmpl-"),
                created=format_timestamp(),
                model=request.model,
                choices=[
                    ChatCompletionChoice(
                        index=0,
                        message=ChatMessage(role="assistant", content=response_text),
                        finish_reason="stop"
                    )
                ],
                usage=ChatCompletionUsage(
                    prompt_tokens=prompt_tokens,
                    completion_tokens=completion_tokens,
                    total_tokens=total_tokens
                )
            )
            
            logger.info(f"聊天完成生成成功 - 耗时: {elapsed_time:.4f}s, tokens: {total_tokens}")
            return response
            
        except Exception as e:
            logger.error(f"生成聊天完成响应失败: {str(e)}", exc_info=True)
            raise HTTPException(status_code=500, detail="生成聊天完成响应失败")

    async def generate_streaming_response(self, request: ChatCompletionRequest):
        """生成流式响应"""
        try:
            # 生成唯一ID
            completion_id = generate_id("chatcmpl-")
            created = format_timestamp()
            
            # 发送初始块
            first_chunk = ChatCompletionChunk(
                id=completion_id,
                created=created,
                model=request.model,
                choices=[
                    ChatCompletionChunkChoice(
                        index=0,
                        delta=ChatCompletionChunkDelta(role="assistant", content=""),
                        finish_reason=None
                    )
                ]
            )
            yield f"data: {json.dumps(first_chunk.model_dump())}\n\n"
            
            # 流式生成回复
            token_count = 0
            async for token in llm.stream_chat(
                messages=request.messages,
                max_new_tokens=request.max_tokens or 150,
                temperature=request.temperature,
                top_p=request.top_p
            ):
                if token:  # 只发送非空token
                    chunk = ChatCompletionChunk(
                        id=completion_id,
                        created=created,
                        model=request.model,
                        choices=[
                            ChatCompletionChunkChoice(
                                index=0,
                                delta=ChatCompletionChunkDelta(content=token),
                                finish_reason=None
                            )
                        ]
                    )
                    yield f"data: {json.dumps(chunk.model_dump())}\n\n"
                    token_count += 1
            
            # 发送结束块
            final_chunk = ChatCompletionChunk(
                id=completion_id,
                created=created,
                model=request.model,
                choices=[
                    ChatCompletionChunkChoice(
                        index=0,
                        delta=ChatCompletionChunkDelta(),
                        finish_reason="stop"
                    )
                ]
            )
            yield f"data: {json.dumps(final_chunk.model_dump())}\n\n"
            
            # 发送结束标记
            yield "data: [DONE]\n\n"
            
            logger.info(f"流式响应生成完成 - tokens: {token_count}")
            
        except Exception as e:
            logger.error(f"生成流式响应失败: {str(e)}", exc_info=True)
            error_chunk = {
                "error": {
                    "type": "server_error",
                    "message": "生成流式响应失败",
                    "param": None,
                    "code": "server_error"
                }
            }
            yield f"data: {json.dumps(error_chunk)}\n\n"
            yield "data: [DONE]\n\n"

    async def generate_stream(self, messages):
        """生成流式回复（用于WebSocket）"""
        try:
            # 流式生成回复
            async for token in llm.stream_chat(messages=messages):
                yield token
                
        except Exception as e:
            logger.error(f"生成流式回复失败: {str(e)}", exc_info=True)
            raise